### PR TITLE
i#6164: Add new file type for bimodal filtered traces

### PR DESCRIPTION
**Pull request by abhinav92003**
Adds OFFLINE_FILE_TYPE_BIMODAL_FILTERED_WARMUP for traces that are filtered before the TRACE_MARKER_TYPE_FILTER_ENDPOINT marker and unfiltered after it.

Adds invariant_checker checks to make sure the new file type and the filter end-point marker are
always present together, or not at all.

Adds a test that the record filter adds the filter end-point marker also for threads that start after
the filtering stop_timestamp. Adds a shard_count member to basic_counts_t::counters_t to make
it easier to write the new test without hard-coding count of threads in the test.

Fixes: #6164

### CODE REVIEW

File: clients/drcachesim/common/trace_entry.h [hunk 1]
```diff
@@ -808,6 +808,14 @@ typedef enum {
      * The included kernel traces are in the Intel® Processor Trace format.
      */
     OFFLINE_FILE_TYPE_KERNEL_SYSCALLS = 0x1000,
+    /**
+     * Partially filtered trace. The initial part upto the
+     * #TRACE_MARKER_TYPE_FILTER_ENDPOINT marker is filtered, and the later part is not.
```
Comment (derekbruening) [inline clients/drcachesim/common/trace_entry.h:813]: Add a note in the _ENDPOINT marker docs that it always goes with this filetype.

**Review by derekbruening — APPROVED** [review]
Maybe should re-review after invariant addition if desired.

File: clients/drcachesim/common/trace_entry.h [hunk 2]
```diff
@@ -808,6 +808,14 @@ typedef enum {
      * The included kernel traces are in the Intel® Processor Trace format.
      */
     OFFLINE_FILE_TYPE_KERNEL_SYSCALLS = 0x1000,
+    /**
+     * Partially filtered trace. The initial part upto the
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/common/trace_entry.h:812]: Done.

File: clients/drcachesim/tools/filter/cache_filter.cpp [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /* **********************************************************
- * Copyright (c) 2022 Google, Inc.  All rights reserved.
+ * Copyright (c) 2022-2023 Google, Inc.  All rights reserved.
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tools/filter/cache_filter.cpp:2]: Oops. There were initially. Removed.

File: clients/drcachesim/common/trace_entry.h [hunk 4]
```diff
@@ -808,6 +808,14 @@ typedef enum {
      * The included kernel traces are in the Intel® Processor Trace format.
      */
     OFFLINE_FILE_TYPE_KERNEL_SYSCALLS = 0x1000,
+    /**
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/common/trace_entry.h:811]: Done.

[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/common/trace_entry.h:811]: Done.

File: clients/drcachesim/common/trace_entry.h (hunk 1 above)
Comment (abhinav92003) [reply clients/drcachesim/common/trace_entry.h:813]: Done.

[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/common/trace_entry.h:813]: Done. Also added a test for the record_filter that if stop_timestamp is earlier than the first timestamp in the trace, we still emit the endpoint marker. Also the drcachesim op_L0_filter_until_instrs also makes sure to emit that marker even for new threads that start after filtering stopped: https://github.com/DynamoRIO/dynamorio/blob/62e9995ecf25df3e1ae34823f47a66b177f9c989/clients/drcachesim/tracer/output.cpp#L1273.

**Comment by abhinav92003** [comment]
Made a few changes to other analysis tools (record_filter and basic_counts), so requesting review again.

File: clients/drcachesim/tests/record_filter_unit_tests.cpp [hunk 5]
```diff
@@ -412,9 +412,14 @@ test_null_filter()
         std::unique_ptr<dynamorio::drmemtrace::record_filter_t::record_filter_func_t>>
         filter_funcs;
     filter_funcs.push_back(std::move(null_filter));
+    // We use a very small stop_timestamp for the record filter. This is to verify that
+    // we emit the TRACE_MARKER_TYPE_FILTER_ENDPOINT marker for each thread even if
+    // starts after the given stop_timestamp. Since the stop_timestamp is so small, all
```
Comment (derekbruening) [inline clients/drcachesim/tests/record_filter_unit_tests.cpp:417]: nit: s/starts/it starts/

File: clients/drcachesim/tools/basic_counts.h [hunk 6]
```diff
@@ -182,6 +182,12 @@ class basic_counts_t : public analysis_tool_t {
         int_least64_t encodings = 0;
         std::unordered_set<uint64_t> unique_pc_addrs;
 
+        // Metadata for the counts. These are not used for the equality, increment,
+        // or decrement operation, and must be set explicitly.
+
+        // Count of shards that were combined to produce the above counts.
+        int_least64_t shard_count = 1;
```
Comment (derekbruening) [inline clients/drcachesim/tools/basic_counts.h:189]: This should be in the operator+=, operator-=, operator== too right?

**Review by derekbruening — APPROVED** [review]

File: clients/drcachesim/tools/basic_counts.h (hunk 6 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/basic_counts.h:189]: I documented above that this (and any future) metadata vars would not be included in those ops. It's not as simple as counts. E.g. we += counters_t not only across shards but across windows too.

File: clients/drcachesim/tests/record_filter_unit_tests.cpp (hunk 5 above)
Comment (abhinav92003) [reply clients/drcachesim/tests/record_filter_unit_tests.cpp:417]: Done.

**Pull request merged**